### PR TITLE
test vrf listener processes old requests on start up

### PR DESCRIPTION
**Pull request by jinhoonbang**
- test vrf listener processes old requests on start up

### CODE REVIEW

File: core/services/vrf/v2/integration_helpers_test.go [hunk 1]
```diff
@@ -1736,3 +1737,145 @@ func testMaliciousConsumer(
 	}
 	require.Equal(t, 1, len(requests))
 }
+
+func testReplayOldRequestsOnStartUp(
```
Comment (ibrajer) [inline core/services/vrf/v2/integration_helpers_test.go:1741]: Just a general question. What is the difference between integration tests here and integration tests using CTF: https://github.com/smartcontractkit/chainlink/tree/develop/integration-tests
Why do we need both?

File: core/services/vrf/v2/integration_helpers_test.go [hunk 2]
```diff
@@ -1736,3 +1737,145 @@ func testMaliciousConsumer(
...
+	vrfkey, err := app.GetKeyStore().VRF().Create()
+	require.NoError(t, err)
+	registerProvingKeyHelper(t, uni, coordinator, vrfkey)
+	keyHash := vrfkey.PublicKey.MustHash()
+
+	encodedVrfKey, err := app.GetKeyStore().VRF().Export(vrfkey.ID(), testutils.Password)
+	require.NoError(t, err)
+
+	// Shut down the node before making the randomness request
+	require.NoError(t, app.Stop())
```
Comment (ibrajer) [inline core/services/vrf/v2/integration_helpers_test.go:1791]: Do we have to check that the node is no longer running after this or is the `Stop()` method blocking until the node is finally stopped?

File: core/services/vrf/v2/integration_helpers_test.go [hunk 3]
```diff
@@ -1736,3 +1737,145 @@ func testMaliciousConsumer(
...
+
+	// Make the first randomness request.
+	numWords := uint32(20)
+	requestID1, _ := requestRandomnessAndAssertRandomWordsRequestedEvent(t, consumerContract, consumer, keyHash, subID, numWords, 500_000, coordinator, uni.backend, nativePayment)
```
Comment (ibrajer) [inline core/services/vrf/v2/integration_helpers_test.go:1795]: So are we testing Scenario 1 described here: https://www.notion.so/chainlink/VRF-LogPoller-Operation-Guide-a0a62f9d76ec4e4e9d04959b87364a44

File: core/services/vrf/v2/integration_helpers_test.go (hunk 1 above)
Comment (jinhoonbang) [reply core/services/vrf/v2/integration_helpers_test.go:1741]: integration tests here (`/` root package) uses simulated backend, which doesn't actually start a real blockchain and gives more control points about blockchain behavior (e.g. when to commit a block). Useful when you need to manipulate certain states about the blockchain. 

QA integration tests (`/integration-tests`) are meant to run against real chains. This is as close it gets to testing in production like environment. Worth noting that QA integration test also has a DEV mode, which starts a local blockchain that runs on its own, which is different from the simulated blockchain above.

File: core/services/vrf/v2/integration_helpers_test.go (hunk 2 above)
Comment (jinhoonbang) [reply core/services/vrf/v2/integration_helpers_test.go:1791]: yes this is blocking.

File: core/services/vrf/v2/integration_helpers_test.go (hunk 3 above)
Comment (jinhoonbang) [reply core/services/vrf/v2/integration_helpers_test.go:1795]: yes correct

**Review by ibrajer — DISMISSED** [review]
LGTM, the test is relatively simple but covers the exact use case.

Perhaps adding another one in the future like this:
* send multiple requests before the node shuts down
* shut down the node
* send another batch of requests while the node is down
* restart the node
* send another batch of new requests
* validate that all three batches were successfully processed and all requests are fulfilled

Not sure if this test (or a variation of it) would be beneficial or not, but the idea here is to hunt down any potential issues with queueing requests and the node processing that queue.

**Comment by jinhoonbang** [comment]
> LGTM, the test is relatively simple but covers the exact use case.
> 
> Perhaps adding another one in the future like this:
> 
> * send multiple requests before the node shuts down
> * shut down the node
> * send another batch of requests while the node is down
> * restart the node
> * send another batch of new requests
> * validate that all three batches were successfully processed and all requests are fulfilled
> 
> Not sure if this test (or a variation of it) would be beneficial or not, but the idea here is to hunt down any potential issues with queueing requests and the node processing that queue.

Thanks for reviewing this! Agreed on the suggested test use case. We discussed a similar test during one of the stand ups and decided to add it in QA integration test. it's not easy to add such tests in the integration tests here because the `Application` instance cannot be re-started once stopped.

**Review by ibrajer — APPROVED** [review]

**Pull request merged**